client/cmd/dexc/ui/depthchart.go

```diff
@@ -0,0 +1,279 @@
+	if width == 0 || height == 0 {
+		return nil
+	}
+	runeRows := make([]string, 0, height)
+	edge, err := interpolate(pts, width, height)
+	if err != nil {
+		// Don't update the UI from a Draw method, so we won't print an error here.
+		// TODO: Add a file-only logger for logging these errors?
+		return nil
+	}
+
+	for iy := 0; iy < height; iy++ {
+		y := height - iy - 1
+		row := make([]rune, 0, width)
+		for _, pt := range edge {
+			switch {
+			case pt.y < y:
+				row = append(row, uniSpace)
+			case pt.y == y:
+				// fmt.Printf("-- drawing at %d\n", y)
+				row = append(row, pt.ch)
+			default:
+				row = append(row, fullBlock)
+			}
+		}
+		runeRows = append(runeRows, string(row))
+	}
+	return runeRows
+}
+
+// Focus is to the tview.Box Focus method, wrapped here to force a redraw of
+// the chart.
+func (c *depthChart) Focus(delegate func(p tview.Primitive)) {
+	c.focus = true
+	// Have to increment the seqID on focus to force redraw.
+	c.mtx.Lock()
+	c.seq++
+	c.mtx.Unlock()
+	// Pass the delegate to the embedded Box's method.
+	c.Box.Focus(delegate)
+}
+
+// Blur is to the tview.Box Focus method. Used to set the focus field.
+func (c *depthChart) Blur() {
+	c.focus = false
+	c.Box.Blur()
+}
+
+// See the unicode block elements at
+// https://en.wikipedia.org/wiki/Box-drawing_character#Unicode
+const (
+	uniSpace          rune = '\u0020'
+	bottomEight       rune = '\u2581'
+	bottomQuarter     rune = '\u2582'
+	bottomThreeEights rune = '\u2583'
+	bottomHalf        rune = '\u2584'
+	bottomFiveEights  rune = '\u2585'
+	bottomTwoThirds   rune = '\u2586'
+	bottomSevenEights rune = '\u2587'
+	fullBlock         rune = '\u2588'
+	// The remaining runes are not used right now, but they may be handy for
+	// additional sub-character detail in the future.
+	// bottomRightHeavy rune = '\u259f'
+	// bottomRightLite  rune = '\u2597'
+	// bottomLeftHeavy  rune = '\u2599'
+	// bottomLeftLite   rune = '\u2596'
+	// topRightHeavy    rune = '\u259c'
+	// topRightLite     rune = '\u259d'
+	// topLeftHeavy     rune = '\u259b'
+	// topLeftLite      rune = '\u2598'
+	// shadedBlock      rune = '\u2593'
+)
+
+// Blur is to the tview.Box Focus method. Used to set the focus field.
+type edgePoint struct {
+	y  int
+	ch rune
+}
+
+// Translate the x,y chart data into a width edgePoints, each edgePoint with
+// 0 < y < height, and an appropriate character.
+//
+// The algorithm projects the line graph onto the width x height grid, and finds
+// the points where the charted line intersects the cell edges. The value
+// assigned to a column is the average of the intersects of the left and
+// right sides.
+func interpolate(pts []depthPoint, width, height int) ([]edgePoint, error) {
+	xMin, xMax, yMin, yMax, err := readLimits(pts)
+	if err != nil {
+		return nil, err
+	}
+	numPts := len(pts)
+	yRange := float64(yMax - yMin)
+	xRange := float64(xMax - xMin)
+	yScaler, xScaler := yRange/float64(height), xRange/float64(width)
+	lastY := float64(pts[0].y) / yScaler
```

Comment:
calcRows won't pass a zero-length pts slice, but perhaps a check and early return in case interpolate sees more general use?
Also with a slice of length 1 it returns an error.  I don't know what an appropriate edge would be in that case, but it seems like a point on the chart?  I wouldn't mind terribly if the depth chart were just empty until at least two orders per side showed up though, as long as the program doesn't do anything bad.